### PR TITLE
Search using charged_quality (such as welding)

### LINKED ISSUE #63094: Can't search/filter using "welding" quality in AIM
### Describe the bug

I can't seem to search for "welding" quality using the Advanced Inventory Manager.

### Steps to reproduce

1.  Have an "arc welder" in your inventory
2. Confirm that it has "2 welding" quality
3. Open the AIM and select your inventory
4. Filter using "q:weld"
5. Returned results are empty

### Expected behavior

1.  Have an "arc welder" in your inventory
2. Confirm that it has "2 welding" quality
3. Open the AIM and select your inventory
4. Filter using "q:weld"
5. Arc Welder should be in the results

### Screenshots

Game Version:

Today I finally decided to upgrade to version 2023-01-08-0614 (windows) in hopes it is fixed but I see it is still there.

### Versions and configuration

Version 2023-01-08-0614 (windows)

Mods loaded:

- no_npc_food
- no_fungal_growth
- package_bionic_professions
- blazeindustries
- magiclysm
- alt_map_key
- my_sweet_cataclysm
- mma
- speedydex
- stats_through_kills
- statsThroughSkills
- translate_dialogue

### Additional context

This is not a new issue so it is most likely a very old bug.

### PR DESCRIPTION
**Pull request by Brambor**
#### Summary
Bugfixes "Charged tools can be searched by quality"

#### Purpose of change
Fix #63094

Also, fix a docs English sentence.

Also added `%` to failure chance. When I saw this number, it was not clear it is meant to be %.
image

*Edit:* And light refactoring.

#### Describe the solution
If filtering by `q:` return true if the quality matches `charged_quality` or `quality`, not just `quality`.

It is using new method defined in `itype.h` ~and `item.h`~:
```cpp
bool has_any_quality( const std::string &quality ) const
```

#### Describe alternatives you've considered
Unrelated code refactoring (duplicate code in `clzones.cpp`).
Splitting into ~3~ 4 PRs.
More extensive testing.

#### Testing
`V`iew menu
1. Spawn welder to visible tile.
2. Go to `V`iew menu.
3. Filter `q:weld`.
4. Observe welder is there.

AIM
1. Spawn welder to adjasent tile.
2. Open AIM (Advanced Inventory Manager).
3. Filter `q:weld`.
4. Observe welder is there.

Crafting Menu
1. Learn welder recipe.
2. Open crafting menu.
3. Search for `q:weld`
4. Observe welder is there.

#### Additional context
*Edit:*
I have removed syntactic sugar of this type:
```
<some_return_type> item::FUNCTION_NAME() const
{
    return type->FUNCTION_THING;  // sometimes call, sometimes value
}
```

for these `FUNCTION_NAME`:

- `integral_length`
- `is_scannable`
- `degrade_increments`
- `quality_of`

I haven't done it for  | the reason
--- | ---
`base_damage_thrown` | it has TODO
`count_by_charges` | too many references
`damage_min` | consistency with `damage()` and `damage_max()`
`damage_max` | consistency with `damage()` and `damage_min()`

### CODE REVIEW

File: src/item.h [hunk 1]
```diff
@@ -1163,6 +1163,10 @@ class item : public visitable
         * The ids of all the qualities this contains.
         */
         const std::map<quality_id, int> &quality_of() const;
+        /**
+        * True if item has given quality or charged_quality (regardless of current charge).
+        */
+        bool has_any_quality( const std::string &quality ) const;
```
Comment (irwiss) [inline src/item.h:1169]: Putting this on item is possibly too much syntax sugar - it doesn't do anything with the item itself and if you have the item then you have direct access to it's type field in 6 characters, this overloads an already giant item class for not much gain considering you can just do `i.type->has_any_quality(...`

imo as a pure ui helper to do string matching this feels like it should just be a static function near the ui call, otoh there's a bunch of these hanging off public interfaces anyway

**Review by anothersimulacrum — APPROVED** [review]
Percent change looks good, thank you.

File: src/item.h (hunk 1 above)
Comment (Brambor) [reply src/item.h:1169]: Thanks for feedback, will do in the evening.

Comment (Brambor) [reply src/item.h]: I have done this and some simple similar cases.
I haven't done it for  | the reason
--- | ---
`base_damage_thrown` | it has TODO
count_by_charges | too many references
damage_min | consistency with `damage()` and `damage_max()`
damage_max | consistency with `damage()` and `damage_min()`

**Comment by Brambor** [comment]
Another test failed on "slow":

[Clang 12, Ubuntu, Tiles, ASan](https://github.com/CleverRaven/Cataclysm-DDA/actions/runs/4467378147/jobs/7857658292?pr=64322#logs)
failed 23 minutes ago in 1h 40m 24s

I closed this & reopened it to restart tests.

**Comment by anothersimulacrum** [comment]
https://github.com/CleverRaven/Cataclysm-DDA/pull/64421 is trying to fix the test failure

**Pull request merged**